<a href="https://colab.research.google.com/github/SANJEEVKUMARBG/Binary-Classification-Logistic-reg.-/blob/main/Rock_or_Mine_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IMPORTING LIBRARIES

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, f1_score

In [2]:
df = pd.read_csv('/kaggle/input/sonar-dataset/Sonar Data.csv', header=None)

In [3]:
df.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

Using SVM might give us the good result because dimentions are more and chance for curse of dimetionality is high.
Followed by boosting algorithems .

In [5]:
df.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [6]:
df[60].value_counts()/len(df)*100

M    53.365385
R    46.634615
Name: 60, dtype: float64

Data has almost equal number of instances. No problem of data imbalance

In [7]:
df.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [8]:
#checking for missing values
df.isnull().sum()
#no null values

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

Train test split

In [9]:
X = df.drop(60, axis=1)
y = df[60]

In [10]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X,y,test_size = 0.25, stratify=y, random_state=12)

In [11]:
print(Xtrain.shape, Xtest.shape, X.shape)

(156, 60) (52, 60) (208, 60)


Model Building

In [12]:
#Logistic Regression

LR = LogisticRegression()
LR.fit(Xtrain, ytrain)
ypred = LR.predict(Xtest)
print(accuracy_score(ytest,ypred))

0.7692307692307693


In [13]:
print(classification_report(ytest,ypred))

              precision    recall  f1-score   support

           M       0.79      0.79      0.79        28
           R       0.75      0.75      0.75        24

    accuracy                           0.77        52
   macro avg       0.77      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



In [14]:
# SVM

from sklearn.svm import SVC
model = SVC()
model.fit(Xtrain,ytrain)
ypred_svm = model.predict(Xtest)
print(accuracy_score(ytest,ypred_svm))

0.7884615384615384


In [15]:
print(classification_report(ytest,ypred_svm))

              precision    recall  f1-score   support

           M       0.74      0.93      0.83        28
           R       0.88      0.62      0.73        24

    accuracy                           0.79        52
   macro avg       0.81      0.78      0.78        52
weighted avg       0.81      0.79      0.78        52



In [16]:
# Lets try with Decision tree and a bagging model

In [17]:
from sklearn.tree import DecisionTreeClassifier
t_model = DecisionTreeClassifier(criterion='entropy')
t_model.fit(Xtrain,ytrain)
ypred_dt = t_model.predict(Xtest)
print('accuracy score',accuracy_score(ytest,ypred_dt), '\n')
print('Classification report','\n',classification_report(ytest,ypred_dt))

accuracy score 0.7692307692307693 

Classification report 
               precision    recall  f1-score   support

           M       0.77      0.82      0.79        28
           R       0.77      0.71      0.74        24

    accuracy                           0.77        52
   macro avg       0.77      0.76      0.77        52
weighted avg       0.77      0.77      0.77        52



In [18]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier
e_model = RandomForestClassifier()
e_model.fit(Xtrain,ytrain)
ypred_rf = e_model.predict(Xtest)
print('accuracy score',accuracy_score(ytest,ypred_rf), '\n')
print('Classification report','\n',classification_report(ytest,ypred_rf))

accuracy score 0.8269230769230769 

Classification report 
               precision    recall  f1-score   support

           M       0.79      0.93      0.85        28
           R       0.89      0.71      0.79        24

    accuracy                           0.83        52
   macro avg       0.84      0.82      0.82        52
weighted avg       0.84      0.83      0.82        52



In [19]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


# Boosting
import xgboost as xgb
boost = xgb.XGBClassifier()
boost.fit(Xtrain,ytrain)
ypred_xg = boost.predict(Xtest)
print('accuracy score',accuracy_score(ytest,ypred_xg), '\n')
print('Classification report','\n',classification_report(ytest,ypred_xg))

Boosting algos gave the highest accuracy.

accuracy score 0.8653846153846154 

Classification report 
               precision    recall  f1-score   support

           M       0.84      0.93      0.88        28
           R       0.90      0.79      0.84        24

    accuracy                           0.87        52
   macro avg       0.87      0.86      0.86        52
weighted avg       0.87      0.87      0.86        52